aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/strategy/CleanupResources.java

```diff
@@ -0,0 +1,129 @@
+import java.util.List;
+
+import static org.jclouds.compute.config.ComputeServiceProperties.TIMEOUT_NODE_SUSPENDED;
+import static org.jclouds.compute.config.ComputeServiceProperties.TIMEOUT_NODE_TERMINATED;
+
+@Singleton
+public class CleanupResources {
+
+   public static final String RESOURCE_TYPE = "securitygroup";
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final ECSComputeServiceApi api;
+   private final Predicate<String> instanceSuspendedPredicate;
+   private final Predicate<String> instanceTerminatedPredicate;
+
+   @Inject
+   public CleanupResources(ECSComputeServiceApi api,
+                           @Named(TIMEOUT_NODE_SUSPENDED) Predicate<String> instanceSuspendedPredicate,
+                           @Named(TIMEOUT_NODE_TERMINATED) Predicate<String> instanceTerminatedPredicate
+   ) {
+      this.api = api;
+      this.instanceSuspendedPredicate = instanceSuspendedPredicate;
+      this.instanceTerminatedPredicate = instanceTerminatedPredicate;
+   }
+
+   public boolean cleanupNode(final String id) {
+      final RegionAndId regionAndId = RegionAndId.fromSlashEncoded(id);
+      List<String> securityGroupIds = getSecurityGroupIdsUsedByNode(regionAndId);
+      String instanceId = regionAndId.id();
+      InstanceStatus instanceStatus = Iterables.tryFind(api.instanceApi().getStatus(regionAndId.regionId()),
+              new InstanceStatusPredicate(instanceId)).orNull();
+      if (instanceStatus == null) return true;
+      if (InstanceStatus.Status.STOPPED != instanceStatus.status()) {
+         logger.debug(">> powering off %s ...", id);
+         api.instanceApi().powerOff(instanceId);
+         instanceSuspendedPredicate.apply(id);
+      }
+      logger.debug(">> destroying %s ...", id);
+      api.instanceApi().delete(instanceId);
+      boolean instanceDeleted = instanceTerminatedPredicate.apply(id);
+
+      for (String securityGroupId : securityGroupIds) {
+         logger.debug(">> destroying security group %s ...", securityGroupId);
+         if (cleanupSecurityGroupIfOrphaned(regionAndId.regionId(), securityGroupId)) {
+            logger.debug(">> security group: (%s) has been deleted.", securityGroupId);
+         } else {
+            logger.warn(">> security group: (%s) has not been deleted.", securityGroupId);
+         }
+      }
+
+      return instanceDeleted;
+   }
+
+   private List<String> getSecurityGroupIdsUsedByNode(RegionAndId regionAndId) {
+      List<String> securityGroupIds = Lists.newArrayList();
+      PaginatedCollection<Instance> instances = api.instanceApi().list(regionAndId.regionId(), ListInstancesOptions.Builder.instanceIds(regionAndId.id()));
+      if (instances.isEmpty()) return securityGroupIds;
+
+      Instance instance = Iterables.get(instances, 0);
+      if (instance != null && !instance.securityGroupIds().isEmpty()) {
+         securityGroupIds = instance.securityGroupIds().values().iterator().next();
```

Comment:
Just one security group? Why not getting the entire set?

Reply:
again it's a list ;)